lib/gui/components/safe-webview/safe-webview.js

```diff
@@ -0,0 +1,231 @@
+    if (this.state.shouldLoad) {
+
+      // Events React is unaware of have to be handled manually
+      this.refs.webview.addEventListener('did-start-loading', this.didStartLoading);
+      this.refs.webview.addEventListener('did-fail-load', this.didFailLoad);
+      this.refs.webview.addEventListener('did-stop-loading', this.didStopLoading);
+      this.refs.webview.addEventListener('did-get-response-details', this.didGetResponseDetails);
+      this.refs.webview.addEventListener('will-navigate', this.willNavigate);
+      this.refs.webview.addEventListener('ipc-message', this.ipcMessage);
+    }
+  }
+
+  /**
+   * @summary Remove the Webview events if there is an element
+   */
+  componentWillUnmount() {
+
+    // There is no element to remove events from if 'shouldLoad' is false.
+    if (this.state.shouldLoad) {
+
+      // Events React is unaware of have to be handled manually
+      this.refs.webview.removeEventListener('did-start-loading', this.didStartLoading);
+      this.refs.webview.removeEventListener('did-fail-load', this.didFailLoad);
+      this.refs.webview.removeEventListener('did-stop-loading', this.didStopLoading);
+      this.refs.webview.removeEventListener('did-get-response-details', this.didGetResponseDetails);
+      this.refs.webview.removeEventListener('will-navigate', this.willNavigate);
+      this.refs.webview.removeEventListener('ipc-message', this.ipcMessage);
+    }
+  }
+
+  /**
+   * @summary Start the loading timeout timer
+   */
+  didStartLoading() {
+    if (this.props.timeout) {
+      this.loadingTimeoutId = setTimeout(() => {
+        this.setState({
+          shouldLoad: false
+        });
+
+      // Milliseconds to wait before cancelling the page-load.
+      }, this.props.timeout);
+    }
+  }
+
+  /**
+   * @summary Set the element state to hidden
+   */
+  didFailLoad() {
+    this.setState({
+      shouldLoad: false
+    });
+  }
+
+  /**
+   * @summary Remove the timeout
+   */
+  didStopLoading() {
+
+    // We have loaded, remove the queued timeout action.
+    clearTimeout(this.loadingTimeoutId);
+  }
+
+  /**
+   * @summary Set the element state depending on the HTTP response code
+   * @param {Event} event - Event object
+   */
+  didGetResponseDetails(event) {
+
+    const {
+      httpResponseCode
+    } = event;
+    const HTTP_OK = 200;
+
+    if (httpResponseCode !== HTTP_OK) {
+      this.setState({
+        shouldLoad: false
+      });
+    }
+
+  }
+
+  /**
+   * @param {Event} event - event object
+   */
+  willNavigate(event) {
+    const url = new URL(event.url);
+
+    if (url.hash === '#external') {
+      this.refs.webview.stop();
+
+      url.hash = '';
+
+      if (url.protocol === 'http:' || url.protocol === 'https:') {
+        electron.shell.openExternal(url.href);
+      }
+    }
+  }
+
+  ipcMessage(event) {
+    const maybeArg = _.get(event.args, 0);
+
+    if (_.get(maybeArg, 'type') === 'error') {
```

Comment:
Can we extract the type and the message as variables given that we're using them twice here?